### PR TITLE
Booking API

### LINKED ISSUE #50: Filter trainer test is not working
- There have been an update with filter trainer api
- The test have not been updated yet
- Now it is commented out

### PR DESCRIPTION
**Pull request by nopparujp**
get-booking return only upcoming appointment
update-booking
delete-booking
create-booking

### CODE REVIEW

**Comment by noppakorn** [comment]
@nopparuj917 Please merge the booking branch into this branch. Something still is not updated.

File: controllers/trainer_controller.go [hunk 1]
```diff
@@ -10,31 +10,30 @@ import (
 	"github.com/gin-gonic/gin"
 )
 
-type FilterTrainerInput struct {
+type FilterTrainerForm struct {
 	Specialty []string `json:"specialty"`
 	Limit     int      `json:"limit" binding:"required"`
 	FeeMin    float64  `json:"feeMin"`
 	FeeMax    float64  `json:"feeMax"`
```
Comment (noppakorn) [inline controllers/trainer_controller.go]: Type is wrong

File: responses/booking_response.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package responses
+
+import "trainder-api/models"
+
+type CreateBookingResponse struct {
+	Status  int    `json:"status"`
+	Message string `json:"message,omitempty"`
+}
+
+type GetBookingsResponse struct {
```
Comment (noppakorn) [inline responses/booking_response.go:10]: Bookings should bev Booking?

File: utils/inits/db_setup.go [hunk 3]
```diff
@@ -21,15 +21,25 @@ func InitializeDatabase() {
...
 		fmt.Println("User collections does not exist, creating `users` collection with `root` user")
 		CreateRootUser()
 	}
+	if !bookingCollectionExists {
+		fmt.Println("Booking collection does not exist, creating `bookings` collection")
+		err := db.CreateCollection(ctx, "bookings")
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
```
Comment (noppakorn) [inline utils/inits/db_setup.go]: why is this check required? there is no need to pre-populate the booking collections? the init db features is used for population of static data only

File: controllers/booking_controller.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+// @Description Delete a booking with the specified bookingId
+// @Tags bookings
+// @Accept json
+// @Produce json
+// @Security BearerAuth
+// @Param json_in_ginContext body DeleteBookingForm true "put DeleteBookingForm details and pass to gin.Context"
+// @Success	200		{object}	responses.DeleteBookingResponse	"Successfully delete booking"
+// @Failure	400		{object}	responses.DeleteBookingResponse	"Bad Request, missing filed of objectId or cannot find bookingObjectId"
+// @Router /protected/delete-booking [post]
+func DeleteBooking() gin.HandlerFunc {
```
Comment (noppakorn) [inline controllers/booking_controller.go:157]: Should this routes be a `DELETE` instead of a `POST`?

File: controllers/trainer_controller.go [hunk 5]
```diff
@@ -10,31 +10,30 @@ import (
...
-	// Rating     float64 `json:"Rating" binding:"required"`
 }
-type UpdateTrainerInput struct {
+type UpdateTrainerDetails struct {
 	Specialty      []string `json:"specialty"`
 	Rating         float64  `json:"rating"`
 	Fee            float64  `json:"fee"`
```
Comment (noppakorn) [inline controllers/trainer_controller.go]: wrong type?

File: controllers/trainer_controller.go [hunk 6]
```diff
@@ -10,31 +10,30 @@ import (
...
 	TraineeCount   int32    `json:"traineeCount"`
```
Comment (noppakorn) [inline controllers/trainer_controller.go]: use `int` instead of `int32` for consistency

File: controllers/trainer_controller.go [hunk 7]
```diff
@@ -228,13 +228,14 @@ func FilterTrainer() gin.HandlerFunc {
...
-// @Param		GetReviewsInput	body		GetReviewsInput 	true	"Parameters for querying trainer reviews"
+// @Param		GetReviewsInput	body		GetReviewsForm true	"Parameters for querying trainer reviews"
 // @Success		200				{object}	responses.TrainerReviewsResponse
+// @Failure		400				{object}	responses.TrainerReviewsResponse
 // @Security	BearerAuth
 // @Router		/protected/reviews [post]
 func GetReviews() gin.HandlerFunc {
 	return func(c *gin.Context) {
-		var input GetReviewsInput
+		var input GetReviewsForm
```
Comment (noppakorn) [inline controllers/trainer_controller.go:215]: A `GET` method would be better.

**Review by noppakorn — CHANGES_REQUESTED** [review]

File: responses/booking_response.go (hunk 2 above)
Comment (looknatc) [reply responses/booking_response.go:10]: it return multiple bookings

File: controllers/trainer_controller.go (hunk 7 above)
Comment (nopparujp) [reply controllers/trainer_controller.go:215]: use this for now

File: controllers/booking_controller.go (hunk 4 above)
Comment (nopparujp) [reply controllers/booking_controller.go:157]: Use this for now

**Review by noppakorn — APPROVED** [review]

**Pull request merged**